vm-setup/roles/v1aX_integration_test/tasks/upgrade.yml

```diff
@@ -147,26 +156,48 @@
       state: directory
       recurse: yes
     with_items:
+    - dev-repository/cluster-api/{{ CAPIRELEASE }}
+    - dev-repository/bootstrap-kubeadm/{{ CAPIRELEASE }}
+    - dev-repository/control-plane-kubeadm/{{ CAPIRELEASE }}
+    - overrides/infrastructure-metal3/{{ CAPIRELEASE }}
     - dev-repository/cluster-api/{{ CAPI_REL_TO_VERSION }}
     - dev-repository/bootstrap-kubeadm/{{ CAPI_REL_TO_VERSION }}
     - dev-repository/control-plane-kubeadm/{{ CAPI_REL_TO_VERSION }}
     - overrides/infrastructure-metal3/{{ CAPM3_REL_TO_VERSION }}
 
+  - name: Copy files
+    vars:
+      working_dir: "{{HOME}}/.cluster-api"
+    copy: src="{{working_dir}}/{{item.src}}" dest="{{working_dir}}/{{item.dest}}"
+    with_items:
+    - {
+        src: "dev-repository/cluster-api/v0.4.99/core-components.yaml",
+        dest: "dev-repository/cluster-api/{{CAPIRELEASE}}/core-components.yaml"
+      }
+    - {
+        src: "dev-repository/bootstrap-kubeadm/v0.4.99/bootstrap-components.yaml",
+        dest: "dev-repository/bootstrap-kubeadm/{{CAPIRELEASE}}/bootstrap-components.yaml"
+      }
+    - {
+        src: "dev-repository/control-plane-kubeadm/v0.4.99/control-plane-components.yaml",
+        dest: "dev-repository/control-plane-kubeadm/{{CAPIRELEASE}}/control-plane-components.yaml"
+      }
+
   - name: Create next version controller CRDs
     vars:
-      working_dir: "{{HOME}}/.cluster-api/"
+      working_dir: "{{HOME}}/.cluster-api"
     copy: src="{{working_dir}}/{{item.src}}" dest="{{working_dir}}/{{item.dest}}"
     with_items:
     - {
-        src: "dev-repository/cluster-api/{{CAPIRELEASE_HARDCODED}}/core-components.yaml",
+        src: "dev-repository/cluster-api/{{CAPIRELEASE}}/core-components.yaml",
         dest: "dev-repository/cluster-api/{{CAPI_REL_TO_VERSION}}/core-components.yaml"
       }
     - {
-        src: "dev-repository/bootstrap-kubeadm/{{CAPIRELEASE_HARDCODED}}/bootstrap-components.yaml",
+        src: "dev-repository/bootstrap-kubeadm/{{CAPIRELEASE}}/bootstrap-components.yaml",
```

Comment:
It looks to me that you are doing the following. 

1. create folder with current_version
2. create folder with next_version
3. copy data from hard_coded to current_version
4. copy data from current_version to next_version

I do not think you need the current version as a transition.